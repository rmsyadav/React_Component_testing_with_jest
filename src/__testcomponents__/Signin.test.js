import { rest } from 'msw';
import renderer from 'react-test-renderer';
import { render, screen,act } from '@testing-library/react';
import { setupServer } from 'msw/node';
import SignIn from '../Components/Signin';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event'

const mockTasks ={users:"Ramashankar Kumar",token:"abdsdhkdhdjs@564783dndjsfaj",statusCode:200,statusMessage:"success!"}
const mockTasks2 ={statusCode:403,statusMessage:"error!"}

const getTasksPath ='/shopping/api/v1/login';

const formHandler = rest.post(getTasksPath, async (req, res, ctx) =>{
  const  body = await req.json();
  const {email,password} = await body;
  if(email === "rmsyadav567@gmail.com" && password === "rmsyadav@567"){
    return  res(ctx.json(mockTasks))
  } else {
    return res(ctx.json(mockTasks2))
  }
  
 }  
);

const mswServer = setupServer(formHandler);



describe('Signin component testing', () => {

   beforeAll(() => mswServer.listen());
   afterEach(() => mswServer.resetHandlers());
   afterAll(() => mswServer.close());
   
   it('Signin component snapshot testing ',()=>{

     const signin = renderer.create(<BrowserRouter><SignIn></SignIn></BrowserRouter>).toJSON();
     expect(signin).toMatchSnapshot();
   })

   it('succesfully loggedin to my website!', async () => {
    const user = userEvent.setup();
     render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
     const textBoxElement = screen.getByTestId("email");
     await act(async()=>{
      await user.type(textBoxElement,'rmsyadav567@gmail.com');
     }) 
     expect(screen.getByTestId('email')).toHaveValue('rmsyadav567@gmail.com');
     const passwordElement = screen.getByTestId("password");
     await act(async()=>{
      await user.type(passwordElement,'rmsyadav@567');
     })
     expect(screen.getByTestId('password')).toHaveValue('rmsyadav@567');
     const submitBtnElement = screen.getByTestId('submitbtn');
     await act(async()=>{
      await user.click(submitBtnElement);
     }) 
     const displayedSuccessMessage = await screen.findByText("Successfully you have loggedIn!");
     expect(displayedSuccessMessage).toBeInTheDocument(); 
   });
   it('Failed to login to my website!', async () => {
    const user = userEvent.setup();
     render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
     const textBoxElement = screen.getByTestId("email");
     await act(async()=>{
      await user.type(textBoxElement,'rmsyadav789@gmail.com');
     })
     expect(screen.getByTestId('email')).toHaveValue('rmsyadav789@gmail.com');
     const passwordElement = screen.getByTestId("password");
     await act(async()=>{
      await user.type(passwordElement,'rmsyadav@537');
     })
     expect(screen.getByTestId('password')).toHaveValue('rmsyadav@537');
     const submitBtnElement = screen.getByTestId('submitbtn');
     await act(async()=>{
      await user.click(submitBtnElement);
     }) 
     const displayedSuccessMessage = await screen.findByText("email or password is wrong!");
     expect(displayedSuccessMessage).toBeInTheDocument(); 
   });

   test('renders the todo component', async() => {
    const user = userEvent.setup();
    render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
    const buttonElement = screen.getByTestId('switchbtn');
    expect(buttonElement).toBeInTheDocument();
    await act(async()=>{
      await user.click(buttonElement);
    })
  });

  it('Check validation error message for email!', async () => {
     const user = userEvent.setup();
     render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
     const textBoxElement = screen.getByTestId("email");
     await act(async()=>{
      await user.type(textBoxElement,'rmsyadav789@gmail.com');
     })
     expect(screen.getByTestId('email')).toHaveValue('rmsyadav789@gmail.com');
     const submitBtnElement = screen.getByTestId('submitbtn');
     await act(async()=>{
      await user.click(submitBtnElement);
     })
     const displayedSuccessMessage = await screen.findByText("Required");
     expect(displayedSuccessMessage).toBeInTheDocument(); 
   });
   it("check error message for invalid email address",async()=>{
     const user = userEvent.setup();
     render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
     const textBoxElement = screen.getByTestId("email");
     await act(async()=>{
      await user.type(textBoxElement,'rmsyadav789gmail');
     })
     const passwordElement = screen.getByTestId("password");
     await act(async()=>{
      await user.type(passwordElement,'rmsyadav@537');
     })
     const submitBtnElement = screen.getByTestId('submitbtn');
     await act(async()=>{
      await user.click(submitBtnElement);
     })
     const displayedSuccessMessage = await screen.findByText("Invalid email address");
     expect(displayedSuccessMessage).toBeInTheDocument(); 
   });
   it('Check validation error message for email!', async () => {
    const user = userEvent.setup();
    render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
     const passwordElement = screen.getByTestId("password");
     await act(async()=>{
      await user.type(passwordElement,'rmsyadav@537');
     })
    const submitBtnElement = screen.getByTestId('submitbtn');
    await act(async()=>{
      await user.click(submitBtnElement);
     })
    const displayedSuccessMessage = await screen.findByText("Required");
    expect(displayedSuccessMessage).toBeInTheDocument(); 
  });
  it('Check validation error message for email and passwor both', async () => {
    const user = userEvent.setup();
    render(<BrowserRouter><SignIn></SignIn></BrowserRouter>);
    const submitBtnElement = screen.getByTestId('submitbtn');
    await act(async()=>{
      await user.click(submitBtnElement);
     })
    const displayedSuccessMessage = await screen.findAllByText("Required");
    expect(displayedSuccessMessage).toHaveLength(2); 
  });

 
 });