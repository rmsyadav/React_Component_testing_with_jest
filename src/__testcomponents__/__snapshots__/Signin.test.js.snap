// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin component testing Signin component snapshot testing  1`] = `
Array [
  <section
    className="container gradient-form rounded-3"
    style={
      Object {
        "backgroundImage": "url('https://mdbcdn.b-cdn.net/img/Photos/new-templates/search-box/img4.webp')",
      }
    }
  >
    <div
      className="container py-5"
    >
      <div
        className="row d-flex justify-content-center align-items-center"
      >
        <div
          className="col-xl-10"
        >
          <div
            className="card rounded-3 text-black"
          >
            <div
              className="row g-0"
            >
              <div
                className="col-lg-6"
              >
                <div
                  className="card-body p-md-5 mx-md-4"
                >
                  <div
                    className="text-center"
                  >
                    <img
                      alt="logo"
                      src="https://mdbcdn.b-cdn.net/img/Photos/new-templates/bootstrap-login-form/lotus.webp"
                      style={
                        Object {
                          "width": "185px",
                        }
                      }
                    />
                    <h4
                      className="mt-1 mb-5 pb-1 clipImage"
                    >
                      We are The Lotus Team
                    </h4>
                  </div>
                  <form
                    onSubmit={[Function]}
                  >
                    <h5
                      className="clipImage2 my-3"
                    >
                      Please login to your account
                    </h5>
                    <div
                      className="form-outline mb-4 input-group"
                    >
                      <span
                        className="btn btn-outline-secondary form-field-mandatory-icon"
                      >
                        <i
                          className="bi bi-envelope-at-fill"
                        />
                      </span>
                      <input
                        className="form-control form-control-md"
                        data-testid="email"
                        id="email"
                        name="email"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="email address"
                        type="email"
                        value=""
                      />
                    </div>
                    <div
                      className="form-outline mb-4 input-group"
                    >
                      <span
                        className="btn btn-outline-secondary form-field-mandatory-icon"
                      >
                        <i
                          className="bi bi-shield-lock-fill"
                        />
                      </span>
                      <input
                        className="form-control form-control-md"
                        data-testid="password"
                        id="password"
                        name="password"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                    </div>
                    <div
                      className="text-center pt-1 mb-5 pb-1"
                    >
                      <input
                        className="form-control btn btn-primary btn-block fa-lg gradient-custom-2 mb-3"
                        data-testid="submitbtn"
                        type="submit"
                        value="Login"
                      />
                    </div>
                    <div
                      className="d-flex align-items-center justify-content-center pb-4"
                    >
                      <p
                        className="mb-0 me-2"
                      >
                        Don't have an account?
                      </p>
                      <button
                        className="btn btn-outline-danger"
                        data-testid="switchbtn"
                        onClick={[Function]}
                        type="button"
                      >
                        Create new
                      </button>
                    </div>
                  </form>
                </div>
              </div>
              <div
                className="col-lg-6 d-flex align-items-center gradient-custom-2"
              >
                <div
                  className="text-white px-3 py-4 p-md-5 mx-md-4"
                >
                  <h4
                    className="mb-4"
                  >
                    We are more than just a company
                  </h4>
                  <p
                    className="large mb-0"
                  >
                    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <footer
    className="container-fluid footer-img rounded-2"
  >
    <div
      className="text-center footer-fontsize-mobile"
    >
      © 2023 Copyright:
      <span
        className="text-reset fw-bold"
      >
         Learning App
      </span>
    </div>
  </footer>,
]
`;
